Add support for 'experimental' colon syntax (public sublibraries)

kczulko opened the pull request.

kczulko commented:
Hi @facundominguez 

Not sure if you'll like that but I wanted to give it a try

Things which I've found cumbersome:
1. Following cabal definition:
```cabal
library
    exposed-modules:
      PackageA.Exposed.A
    other-modules:
      PackageA.Other.B
      PackageA.Other.D
      PackageA.Other.E
    build-depends:
      base,
      inspection-testing,
      package-b:sublib,
      package-b
```
returns following dependencies list:
```haskell
[
  Dependency (PackageName \"base\") AnyVersion (fromList [LMainLibName]),
  Dependency (PackageName \"inspection-testing\") AnyVersion (fromList [LMainLibName]),
  Dependency (PackageName \"package-b\") (IntersectVersionRanges AnyVersion AnyVersion) (
    fromList [LMainLibName, LSubLibName (UnqualComponentName \"sublib\")]
  )
]
```
so it looks like cabal (at least 3.2 which is taken here from ghc 8.10.4) groups everything referring to package-b into one `Dependency` object (four deps defined in cabal file but only three `Dependency` values)
2. I decided to pass <packageName>:<sublibName> from `cabalscan` to golang component and I left `internal-library` prefix untouched. That's because it allows me to keep cabal's 'shadowing' feature working (well... this one is causing bit of headache on the cabal side after introducing public sublibraries... at least this is what I've found on their (cabal) github issues).

facundominguez commented on gazelle_cabal/dependency_resolution.go:
```diff
@@ -206,14 +206,26 @@ func getPluginLabel(
 
+// Handle experimental colon syntax in cabal
+// (e.g. public internal library reference: package:sublibName)
+func resolvePackageName(packageName string) (string, string) {
```
I'd rename the function to `splitDependencyString`. And would document it as:
```go
// Splits a dependency string into a package name and a library name
func splitDependencyString(dep string) (string, string) {
```

facundominguez commented on gazelle_cabal/dependency_resolution.go:
```diff
@@ -224,8 +236,10 @@ func getPackageLabel(
+
+		// indeed internal library can be refereced form the package of its definition
+		// or it can be referenced using experimental colon syntax (a.k.a public internal libraries)
+		if (r.Label.Repo == from.Repo && r.Label.Pkg == from.Pkg) || (len(parentPackage) > 0) {
```
Note the typos in the comment.

Still, is this the right behavior? If a dependency string refers to a sublibrary of another package, this code will chose instead an internal library.

kczulko replied:
I probably don't get your question :/

The problem I see now is that I must check value of `parentPackage`. That's because there can be a situation where e.g. two or more packages will define a sublibrary with the same name, e.g. `package-a:sublib`, `package-b:sublib`, `package-c:sublib`. My code here if unfortunately ignoring left hand side of the colon syntax so `package-a:sublib` can be resolved with this code to `//package-b:sublib`

The first part of this `if` statement is responsible for `shadowing` feature - if we detect a dependency which resolves to some 'local definition' then, we get this local definition from `internal_library` pool rather then from `@stackage` repository.

Second part takes a dependency from `internal_library` pool only if it was found there and if it was referenced with colon syntax - but for sure I need to fix ambiguity case there.

facundominguez replied:
I think you get it.

The case I was referring to is, given that you are resolving dependencies of `package-a`, and that one such dependency is `package-b:sublib`, if `package-a` contains an internal library called `sublib`, this code will chose the internal library instead of the intended sub-library.

facundominguez reviewed (commented):
Thanks @kczulko! There are some things to revise still.

facundominguez commented:
Since dependencies can contain colons now, the `Imports` callback will have to set a non-empty prefix in the default case in order to avoid ambiguity.

facundominguez commented on cabalscan/src/CabalScan/RuleGenerator.hs:
```diff
@@ -259,8 +260,18 @@ findModulePaths componentName cabalFilePath hsSourceDirs moduleNames = do
+          identifierOf (Cabal.LSubLibName name) = pkgName <> ":" <> Text.pack (Cabal.unUnqualComponentName name)
+          identifierOf _ = pkgName
+        in
+          map identifierOf $ toList $ Cabal.depLibraries dep
```
I'd qualify this use of `toList` as `Set.toList` to save the trouble of visiting the import list to guess what it does.

kczulko commented:
@facundominguez ,

I tried to address issues you've found. I also refactored a bit function responsible for dependency resolution. I think now it's clear what is the resolution order and the logic behind it. Please let me know what you think about this. It's still missing proper test cases which I'll add in upcoming days.

Best regards!
Karol

facundominguez commented on gazelle_cabal/dependency_resolution.go:
```diff
@@ -224,42 +223,85 @@ func getPackageLabel(
+    return label.Label{}, fmt.Errorf("Regular library '%s' not found", pkgName)
+}
+
+func resolveAgainstRepository(
```
`grabFromRepository`?

facundominguez commented on gazelle_cabal/dependency_resolution.go:
```diff
@@ -224,42 +223,85 @@ func getPackageLabel(
+	return label.Label{}, fmt.Errorf("Internal library '%s' not found", pkgName)
+}
+
+func resolveAgainstPublicLibraries(
```
`searchInPublicLibraries`?

facundominguez commented on gazelle_cabal/dependency_resolution.go:
```diff
@@ -224,42 +223,85 @@ func getPackageLabel(
+	return label.Label{}
+}
+
+func resolveAgainstPrivateLibraries(
```
`searchInPrivateLibraries`?

facundominguez commented on gazelle_cabal/dependency_resolution.go:
```diff
@@ -224,42 +223,85 @@ func getPackageLabel(
-		// or it can be referenced using experimental colon syntax (a.k.a public internal libraries)
-		if (r.Label.Repo == from.Repo && r.Label.Pkg == from.Pkg) || (len(parentPackage) > 0) {
-			return rel(r.Label, from)
+		if (r.Label.Repo == from.Repo && r.Label.Pkg == from.Pkg) {
```
The comment in line 266 needs editing.

kczulko replied:
This line (and `if` statement) was removed in the latest commit.

facundominguez commented on gazelle_cabal/lang.go:
```diff
@@ -140,8 +140,10 @@ func (*gazelleCabalLang) Imports(c *config.Config, r *rule.Rule, f *rule.File) [
+	    if r.PrivateAttr("visibility") == "private" {
+			prefix = "private_library:"
+		} else {
+			prefix = "public_library:"
```
Indentation characters must be using a mix of tabs and spaces.

facundominguez commented on gazelle_cabal/dependency_resolution.go:
```diff
@@ -224,42 +223,85 @@ func getPackageLabel(
+) (label.Label, error) {
+	parentPackage, depName := splitDependencyString(pkgName)
+
+	if len(parentPackage) > 0 && (parentPackage != depName) {
```
Why should `parentPackage` be `depName`? That appears to be the only way in which the `if` statement can be skipped.

I'd say the library is internal iff `parentPackage` is the same as the name of the rule for which dependencies are being resolved.

Does cabal accept an internal library to be referred as `package-a:lib` from other artifacts of `package-a`?

facundominguez commented on gazelle_cabal/dependency_resolution.go:
```diff
@@ -224,42 +223,85 @@ func getPackageLabel(
+		if (err == nil) {
+			return lbl
+		}
+	}
```
How about unrolling the loop and calling the three functions explicitly?

There is even a fitting `goto` statement in go if you don't want to guard the calls with `if` statements.
https://golang.org/ref/spec#Goto_statements

facundominguez reviewed (commented):
I'd expect the search routines to be simpler if the rules where indexed as
```
public_library:package-name:lib-name
private_library:package-name:lib-name
```

This might require propagating the package name of libraries from cabalscan (same as the visibility).

Another ambiguity in the current indexing scheme arises if two cabal files define sublibraries with the same names. IIUC, only the library name ends up in the index key.

kczulko commented:
Hi Facundo,

Thank you very much for such comprehensive review!

> I'd expect the search routines to be simpler if the rules where indexed as
> 
> ```
> public_library:package-name:lib-name
> private_library:package-name:lib-name
> ```
> 
> This might require propagating the package name of libraries from cabalscan (same as the visibility).

Ok, but what in case of 'stackage' libraries, e.g. `base`? It should be represented in the cabalscan as `base:base`?

> Another ambiguity in the current indexing scheme arises if two cabal files define sublibraries with the same names. IIUC, only the library name ends up in the index key.

I think it won't cause problems, because if statement checks package membership of the dependency it has found:
`r.Label.Pkg == from.Pkg`

> Why should parentPackage be depName? That appears to be the only way in which the if statement can be skipped.
>
> I'd say the library is internal iff parentPackage is the same as the name of the rule for which dependencies are being resolved.
>
> Does cabal accept an internal library to be referred as package-a:lib from other artifacts of package-a?

1. I wanted to exit fast in the following case -> package-a refers to package-b's public named library `package-b:sublib`. Having this 'colon prefix' means that named library from `package-b` must be public, so there is no need run the rest of the code (resolvingAgainstPrivateLibraries). But...
2. You're right... Cabal allows to write within the `package-b` internal lib dependency in the following way: `package-b:sublib`. I missed that so my code doesn't work there :/.

> How about unrolling the loop and calling the three functions explicitly?

Sure, will check that.

Once again huge thanks!

Best regards,
Karol

facundominguez commented:
> Ok, but what in case of 'stackage' libraries, e.g. base? It should be represented in the cabalscan as base:base?

I think I don't understand the question. `stackage` libraries won't end up in the index because they have no haskell rule defining them in the local repo.

> I think it won't cause problems, because if statement checks package membership of the dependency it has found:

I think you would be right if we assumed that there can be at most one cabal file per folder. This is an assumption that we might have introduced when adding support for internal libraries. Perhaps there's some way to remove it with better indexing, otherwise we probably should be checking that we don't get multiple cabal packages in the same folder.

kczulko commented:
> I think I don't understand the question. `stackage` libraries won't end up in the index because they have no haskell rule defining them in the local repo. 

I think I get it now, but please acknowledge my understanding: So.. the cabalscan output should add package prefix to the name of the library rule (and only library rule for now?)? Then yes, you're right, `base` name which has not been defined within any cabal file will be just `base`.

>I think you would be right if we assumed that there can be at most one cabal file per folder. This is an assumption that we might have introduced when adding support for internal libraries. Perhaps there's some way to remove it with better indexing, otherwise we probably should be checking that we don't get multiple cabal packages in the same folder.

It turns out to be a good assumption (at least for now...)
```
$ cd example
$ cp package-b/package-b.cabal package-b/package-c.cabal
$ sed -i -e 's/name:               package-b/name:               package-c/g' package-b/package-c.cabal
$ cabal build all
When using configuration(s) from ***gazelle_cabal/example/cabal.project, the following errors occurred:
The package directory 'package-b' contains multiple .cabal files (which is not currently supported).
$ cabal --version
cabal-install version 3.6.2.0
compiled using version 3.6.2.0 of the Cabal library
```

kczulko commented:
Hi Facundo,

I've implemented some solution with proper prefixing approach. It took me while to find out the best code shape and all possible ways user can put a dependency in cabal (and handle each of them). golang's `goto` stmt didn't help much in this case. I've find out that searching scopes depend on the presence of colon syntax (they are different for `package-b:package-b` and `package-b`). 

Those changes are still without tests but I added new package to `example` directory and this time `package-a.cabal` contains all the build-depends madness that one can imagine. I hope I've covered all the cases there. On the first glance it works as expected. I'll add buildkite checks soon.

Best regards,
Karol

facundominguez commented on gazelle_cabal/dependency_resolution.go:
```diff
@@ -223,85 +214,74 @@ func getPackageLabel(
+			// there is prefix which can reference the public or internal library
+	        fmt.Sprintf(format, publicPrefix, cabalPackageId, libraryId),
+			// TODO: potentially throw error if from.Pkg != cabalPackageId?
+			fmt.Sprintf(format, privatePrefix, from.Pkg, libraryId),
```
If the dependency string contains a packageId (say "tasty:tasty"), it doesn't seem correct to discard it and look for an internal library instead ("package-a:tasty"). Is this code still correct?

kczulko replied:
Ahh I think I know what you mean now... I'm not 100% familiar with cabal but I see that it is allowed define the import for some `respository` dependency in such way (with colon syntax as you did: `tasty:tasty`). In such case we should take tasty from repository, right? I'll have to do something with it. So you're right - this code doesn't cover all the cases.

But just to explain what it does right now. Assuming:
package-a.cabal
```
library sublib
 ...

library
  build-depends:
    package-b:sublib
```
package-b.cabal
```
library sublib:
  visibility:public
```
Order matters here, therefore on the very beginning I'm searching for *public* libraries instead of private ones. If the order would be opposite, than `package-b:sublib` in package-a.cabal would be resolved by us to package-a:sublib.

facundominguez commented on gazelle_cabal/dependency_resolution.go:
```diff
@@ -223,85 +214,74 @@ func getPackageLabel(
+	return []string{
+		// it can be localally defined library which can be either private or public
+		fmt.Sprintf(format, privatePrefix, from.Pkg, depName),
+		fmt.Sprintf(format, publicPrefix, from.Pkg, depName),
```
I think every time this function writes `from.Pkg`, it is really meant `r.PrivateAttr("pkgName")` where `r` is the rule parameter from `setDepsAndPluginsAttributes`.

The package of a Bazel label is not usually a valid Haskell package.

kczulko replied:
Ok, you're right. Will fix that.

facundominguez commented on gazelle_cabal/dependency_resolution.go:
```diff
@@ -211,41 +211,77 @@ func getPluginLabel(
+	return grabFromRepository(packageRepo, pkgName, from)
+}
+
+// Dependency string may contain 'experimental' colon syntax e.g package-b:sublib
+func obtainSearchScopes(depName string, from label.Label) []string {
```
```go
// Produces the key to search the label of a library (either internal or public).
func getRuleIndexKey(depName string, from label.Label) []string {
```

facundominguez commented on gazelle_cabal/dependency_resolution.go:
```diff
@@ -211,41 +211,77 @@ func getPluginLabel(
+	return label.Label{}, fmt.Errorf("Library '%s' referenced from '%s' not found, %s", pkgName, from, importId)
+}
+
+func grabFromRepository(
```
Perhaps it is worth inlining this definition as it is a one-liner.

facundominguez reviewed (commented):
:clap: No more bugs I can point at.

I suggested some renamings and grammar edits

kczulko commented:
Just pointing it here that due to https://github.com/haskell/cabal/issues/7270 the case with hackage/stackage published public sublibrary doesn't exist.

facundominguez reviewed (approved):
Amazing @kczulko! Ready to go!
